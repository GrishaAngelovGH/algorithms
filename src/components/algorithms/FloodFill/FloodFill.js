import { useState, useEffect } from 'react'

import Button from 'react-bootstrap/Button'
import ButtonGroup from 'react-bootstrap/ButtonGroup'

import WaveBackground from '../../WaveBackground'

const FloodFill = () => {
    const grid = [
        ['#', '#', '#', '#', '#', '#', '#', '#', '#'],
        ['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
        ['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
        ['#', ' ', ' ', '#', ' ', ' ', ' ', ' ', '#'],
        ['#', '#', '#', ' ', ' ', ' ', '#', '#', '#'],
        ['#', ' ', ' ', ' ', ' ', '#', ' ', ' ', '#'],
        ['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
        ['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
        ['#', '#', '#', '#', '#', '#', '#', '#', '#'],
    ]

    const generateEmptyGrid = () => new Array(9).fill(0).map(() => new Array(9).fill(false))

    const [currentIndex, setCurrentIndex] = useState(0)
    const [visitedIndexes, setVisitedIndexes] = useState([])
    const [visitedGrid, setVisitedGrid] = useState(generateEmptyGrid())

    let visitedPositions = generateEmptyGrid()

    const nextStep = (i, j) => {
        if (visitedPositions[i][j] || grid[i][j] === '#') return

        visitedPositions[i][j] = true

        setVisitedIndexes(prevState => [...prevState, [i, j]])

        nextStep(i, j)
        nextStep(i + 1, j)
        nextStep(i - 1, j)
        nextStep(i, j + 1)
        nextStep(i, j - 1)
    }

    const handleStep = () => {
        const coords = visitedIndexes[currentIndex]

        if (coords) {
            const newVisitedGrid = [...visitedGrid]

            newVisitedGrid[coords[0]][coords[1]] = true

            setVisitedGrid(newVisitedGrid)

            setCurrentIndex(currentIndex + 1)
        }
    }

    const handleReset = () => {
        setCurrentIndex(0)

        setVisitedIndexes([])

        setVisitedGrid(generateEmptyGrid())

        visitedPositions = generateEmptyGrid()

        nextStep(4, 4)
    }

    useEffect(() => {
        nextStep(4, 4)
        // eslint-disable-next-line
    }, [])

    return (
        <WaveBackground>
            <div className="row g-0 justify-content-center text-center p-3">
                <div className="col-md-12">
                    <h3 className='text-white'>Flood Fill Algorithm</h3>

                    <div className="row justify-content-center text-white">
                        <div className='col-md-6'>
                            {
                                grid.map((row, rowIndex) => (
                                    <div key={rowIndex} className="row justify-content-center">
                                        {
                                            row.map((v, colIndex) => {

                                                if (visitedGrid[rowIndex][colIndex]) {
                                                    return (
                                                        <div key={colIndex} className='col-md-1 border bg-success' style={{ height: 26 }}></div>
                                                    )
                                                }

                                                if (v === '#') {
                                                    return (
                                                        <div key={colIndex} className='col-md-1 border bg-secondary' style={{ height: 26 }}></div>
                                                    )
                                                }

                                                return (
                                                    <div key={colIndex} className='col-md-1 border'>{v}</div>
                                                )
                                            })
                                        }
                                    </div>
                                ))
                            }
                        </div>
                    </div>

                    <div className="row justify-content-center m-3">
                        <div className="col-md-2">
                            <ButtonGroup size='lg' className='mb-2'>
                                <Button onClick={handleStep}>Step</Button>
                                <Button onClick={handleReset}>Reset</Button>
                            </ButtonGroup>
                        </div>
                    </div>

                    <div className="row justify-content-center m-3">
                        <div className="col-md-6">
                            <h4 className='text-white'>Flood fill is an algorithm mainly used to determine a bounded area connected to a given node in a multi-dimensional array. It is a close concept to the bucket tool in paint programs.</h4>
                        </div>
                    </div>
                </div>
            </div>
        </WaveBackground>
    )
}

export default FloodFill