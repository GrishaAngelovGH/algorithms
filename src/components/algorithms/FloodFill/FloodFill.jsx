import { useState, useEffect, useCallback, useMemo, useRef } from 'react'

import Button from 'react-bootstrap/Button'
import ButtonGroup from 'react-bootstrap/ButtonGroup'

import AlgorithmDescription from 'components/AlgorithmDescription'
import WaveBackground from 'components/WaveBackground'

const FloodFill = () => {
	const grid = useMemo(() => [
		['#', '#', '#', '#', '#', '#', '#', '#', '#'],
		['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
		['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
		['#', ' ', ' ', '#', ' ', ' ', ' ', ' ', '#'],
		['#', '#', '#', ' ', ' ', ' ', '#', '#', '#'],
		['#', ' ', ' ', ' ', ' ', '#', ' ', ' ', '#'],
		['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
		['#', ' ', ' ', ' ', '#', ' ', ' ', ' ', '#'],
		['#', '#', '#', '#', '#', '#', '#', '#', '#'],
	], [])

	const generateEmptyGrid = () => new Array(9).fill(0).map(() => new Array(9).fill(false))

	const [currentIndex, setCurrentIndex] = useState(0)
	const [visitedIndexes, setVisitedIndexes] = useState([])
	const [visitedGrid, setVisitedGrid] = useState(generateEmptyGrid())

	const visitedPositions = useRef(generateEmptyGrid())

	const nextStep = useCallback((i, j) => {
		if (visitedPositions.current[i][j] || grid[i][j] === '#') return

		visitedPositions.current[i][j] = true

		setVisitedIndexes(indexes => [...indexes, [i, j]])

		nextStep(i, j)
		nextStep(i + 1, j)
		nextStep(i - 1, j)
		nextStep(i, j + 1)
		nextStep(i, j - 1)
	}, [grid])

	const handleStep = () => {
		const coords = visitedIndexes[currentIndex]

		if (coords) {
			const newVisitedGrid = [...visitedGrid]

			newVisitedGrid[coords[0]][coords[1]] = true

			setVisitedGrid(newVisitedGrid)

			setCurrentIndex(currentIndex + 1)
		}
	}

	const handleReset = () => {
		setCurrentIndex(0)

		setVisitedIndexes([])

		setVisitedGrid(generateEmptyGrid())

		visitedPositions.current = generateEmptyGrid()

		nextStep(4, 4)
	}

	useEffect(() => {
		nextStep(4, 4)
	}, [nextStep])

	return (
		<WaveBackground>
			<div className='row g-0 justify-content-center text-center p-3'>
				<div className='col-md-12'>
					<h3 className='text-white'>Flood Fill Algorithm</h3>

					<div className='row justify-content-center text-white'>
						<div className='col-md-6'>
							{
								grid.map((row, rowIndex) => (
									<div key={rowIndex} className='row justify-content-center'>
										{
											row.map((v, colIndex) => {
												const bgClass = visitedGrid[rowIndex][colIndex] ? 'bg-light' : v === '#' ? 'bg-primary' : ''

												return (
													<div key={colIndex} className={`col-md-1 border ${bgClass}`} style={{ height: 25 }}></div>
												)
											})
										}
									</div>
								))
							}
						</div>
					</div>

					<div className='row justify-content-center m-3'>
						<div className='col-md-2'>
							<ButtonGroup size='lg' className='mb-2'>
								<Button onClick={handleStep}>Step</Button>
								<Button onClick={handleReset}>Reset</Button>
							</ButtonGroup>
						</div>
					</div>

					<div className='row justify-content-center m-3'>
						<div className='col-md-8'>
							<AlgorithmDescription>
								Flood fill is an algorithm mainly used to determine a bounded area connected
								to a given node in a multi-dimensional array.
								It is a close concept to the bucket tool in paint programs.
							</AlgorithmDescription>
						</div>
					</div>
				</div>
			</div>
		</WaveBackground>
	)
}

export default FloodFill