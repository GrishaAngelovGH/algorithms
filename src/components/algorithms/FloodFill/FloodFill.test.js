import { render } from '@testing-library/react'

import FloodFill from './FloodFill'

test('should render component', () => {
    const container = render(<FloodFill />)

    expect(container).toMatchSnapshot()
})