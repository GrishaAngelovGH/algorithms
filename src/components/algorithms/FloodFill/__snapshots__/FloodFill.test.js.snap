// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="waveWrapper waveAnimation"
      >
        <div
          class="waveWrapperInner bgTop"
        >
          <div
            class="wave waveTop"
          />
        </div>
        <div
          class="waveWrapperInner bgMiddle"
        >
          <div
            class="wave waveMiddle"
          />
        </div>
        <div
          class="waveWrapperInner bgBottom"
        >
          <div
            class="wave waveBottom"
          />
        </div>
      </div>
      <div
        class="wave-content"
      >
        <div
          class="row g-0 justify-content-center text-center p-3"
        >
          <div
            class="col-md-12"
          >
            <h3
              class="text-white"
            >
              Flood Fill Algorithm
            </h3>
            <div
              class="row justify-content-center text-white"
            >
              <div
                class="col-md-6"
              >
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border "
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
                <div
                  class="row justify-content-center"
                >
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                  <div
                    class="col-md-1 border bg-primary"
                    style="height: 25px;"
                  />
                </div>
              </div>
            </div>
            <div
              class="row justify-content-center m-3"
            >
              <div
                class="col-md-2"
              >
                <div
                  class="mb-2 btn-group btn-group-lg"
                  role="group"
                >
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Step
                  </button>
                  <button
                    class="btn btn-primary"
                    type="button"
                  >
                    Reset
                  </button>
                </div>
              </div>
            </div>
            <div
              class="row justify-content-center m-3"
            >
              <div
                class="col-md-8"
              >
                <div
                  aria-atomic="true"
                  aria-live="assertive"
                  class="fade toast w-100 show"
                  role="alert"
                >
                  <div
                    class="toast-header text-bg-light"
                  >
                    <h5
                      class="m-0"
                    >
                      Algorithm Description
                    </h5>
                  </div>
                  <div
                    class="toast-body"
                  >
                    <h5
                      class="m-0"
                    >
                      Flood fill is an algorithm mainly used to determine a bounded area connected to a given node in a multi-dimensional array. It is a close concept to the bucket tool in paint programs.
                    </h5>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="waveWrapper waveAnimation"
    >
      <div
        class="waveWrapperInner bgTop"
      >
        <div
          class="wave waveTop"
        />
      </div>
      <div
        class="waveWrapperInner bgMiddle"
      >
        <div
          class="wave waveMiddle"
        />
      </div>
      <div
        class="waveWrapperInner bgBottom"
      >
        <div
          class="wave waveBottom"
        />
      </div>
    </div>
    <div
      class="wave-content"
    >
      <div
        class="row g-0 justify-content-center text-center p-3"
      >
        <div
          class="col-md-12"
        >
          <h3
            class="text-white"
          >
            Flood Fill Algorithm
          </h3>
          <div
            class="row justify-content-center text-white"
          >
            <div
              class="col-md-6"
            >
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border "
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
              <div
                class="row justify-content-center"
              >
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
                <div
                  class="col-md-1 border bg-primary"
                  style="height: 25px;"
                />
              </div>
            </div>
          </div>
          <div
            class="row justify-content-center m-3"
          >
            <div
              class="col-md-2"
            >
              <div
                class="mb-2 btn-group btn-group-lg"
                role="group"
              >
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Step
                </button>
                <button
                  class="btn btn-primary"
                  type="button"
                >
                  Reset
                </button>
              </div>
            </div>
          </div>
          <div
            class="row justify-content-center m-3"
          >
            <div
              class="col-md-8"
            >
              <div
                aria-atomic="true"
                aria-live="assertive"
                class="fade toast w-100 show"
                role="alert"
              >
                <div
                  class="toast-header text-bg-light"
                >
                  <h5
                    class="m-0"
                  >
                    Algorithm Description
                  </h5>
                </div>
                <div
                  class="toast-body"
                >
                  <h5
                    class="m-0"
                  >
                    Flood fill is an algorithm mainly used to determine a bounded area connected to a given node in a multi-dimensional array. It is a close concept to the bucket tool in paint programs.
                  </h5>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
